### PR TITLE
486 - Add ids data grid filters

### LINKED ISSUE #486: Datagrid: Add Filter Row Functionality
Add a filter option to datagrid https://main-enterprise.demo.design.infor.com/components/datagrid/example-filter.html

**Possible Internal Markup**
```js
<ids-data-grid filter-row="true"></ids-data-grid>
```

**Phase 1 MVP**
- Create an ids-filter-column (or similarly named component) and surface this in the filter row
- keyboard should work the same including `filterWhenTyping` option
- use the corresponding components for each filter type
- consider paging examples and add filter row there
- fire filter events
- use menu button for the filter types (icons should be there)
- optional slot `<slot name="${column.id}"></slot>` 

```
<ids-data-grid-filter slot="book-id">
    <lm-code-block></lm-code-block>
 </ids-data-grid-filter>
 ```
  
**Design**
https://main-enterprise.demo.design.infor.com/components/datagrid/example-filter.html
https://www.figma.com/file/4CtBGz8dwvSa6JsBcdtVZQ/?node-id=992%3A2917

**Other**
Pass the following checklist ordered in priority (some may be left for later) https://github.com/infor-design/enterprise-wc/blob/main/doc/CHECKLIST.md

### PR DESCRIPTION
**Pull request by deep7102**
**Explain the details for making this change. What existing problem does the pull request solve?**
Added Ids Data Grid Filters

**Related github/jira issue (required)**:
Closes #486
Related #565 
Related #623 

**Steps necessary to review your pull request (required)**:
- Pull this branch and build/run the demo app
- Navigate to:
- http://localhost:4300/ids-data-grid/filter.html
- http://localhost:4300/ids-data-grid/filter-books.html
- http://localhost:4300/ids-data-grid/filter-custom.html
- See editor all variations and uplift styles
- Added e2e and functional tests
- See the docs

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
- [x] A note to the change log.

**There are some known issues will handle separately**
- Scrolling, align position and bleeding (need to find the way can popup outside from data-grid): date-picker, time-picker, dropdown, menu-button [#565](https://github.com/infor-design/enterprise-wc/issues/565)
- Time picker, dropdown/s not able to open [#565](https://github.com/infor-design/enterprise-wc/issues/565)
- Date picker, year/month popup not able to open [#565](https://github.com/infor-design/enterprise-wc/issues/565)
- Cannot reset to blank value after select one: date-picker, time-picker [#623](https://github.com/infor-design/enterprise-wc/issues/623)

### CODE REVIEW

File: src/assets/data/books.json [hunk 1]
```diff
@@ -7,6 +7,7 @@
 		"transactionCurrency": "Book",
 		"postHistory": "No",
 		"active": "Yes",
+    "inStock": true,
```
Comment (tmcconechy) [inline src/assets/data/books.json:10]: Weird formatting?

File: src/components/ids-data-grid/README.md [hunk 2]
```diff
@@ -154,6 +154,365 @@ When used as an attribute the settings are kebab case, when used in the JS they
...
 - `rowactivated` Fires for each row that is activated.
 - `rowdeactivated` Fires for each row that is deactivated.
+- `filtered` Fires after a filter action occurs, clear or apply filter condition.
+- `filteroperatorchanged` Fires once a filter operator changed.
+- `openfilterrow` Fires after the filter row is opened by the user.
+- `closefilterrow` Fires after the filter row is closed by the user.
+
+## Filters
+
+Data rows can be easily filter based on one or several criteria. Whole filter row can turn on/off by api setting `filterable` and can be disabled by api setting `filter-row-disabled`. The filter conditions apply thru the UI or programmatically. Each column can have its own filter type and turn on/off by columns setting.
```
Comment (tmcconechy) [inline src/components/ids-data-grid/README.md:164]:
```suggestion
Data rows can be filter based on one or several criteria. Whole filter row can turned on/off by the api setting `filterable` and can be disabled by the api setting `filter-row-disabled`. The filter conditions can be applied thru the UI or programmatically. Each column can have its own filter type and turn on/off by columns setting.
```

**Review by tmcconechy — CHANGES_REQUESTED** [review]
Some suggestions/comments, we can make a new issue for the things we know are not working

http://localhost:4300/ids-data-grid/filter.html
- in the Color column type "Gree" -> notice when the filter is applied the focus is lost. Maybe we can only rerender the body (there is a function for that) keeping the focus (or restore it)
- the borders are a different color on https://main-enterprise.demo.design.infor.com/components/datagrid/example-filter.html
- Noting: will need to add column alignment (left, right center)
- the input box in the pager is really big

File: src/assets/data/books.json (hunk 1 above)
Comment (deep7102) [reply src/assets/data/books.json:10]: updated now

**Comment by deep7102** [comment]
@tmcconechy Update now, the point you found.

**Review by tmcconechy — APPROVED** [review]
This is good for me for now. Nice job @deep7102 
Next we should fix the popup issues..

**Review by danielortiz1982 — APPROVED** [review]
Nice work on this PR @deep7102 👍

**Pull request merged**